### PR TITLE
Feature/baryons

### PR DESCRIPTION
**Pull request by felixerben**
Change from last pull request: 
- baryons are now specified by a list of quark-structures, so that also e.g. the lambda baryon can be computed with a single hadrons module.

### CODE REVIEW

File: Grid/qcd/utils/BaryonUtils.h [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+						 const int parity,
+						 const std::vector<int> &wick_contraction,
+						 robj &result)
+{
+
+  Gamma g4(Gamma::Algebra::GammaT); //needed for parity P_\pm = 0.5*(1 \pm \gamma_4)
+
+
+  static const int epsilon[6][3] = {{0,1,2},{1,2,0},{2,0,1},{0,2,1},{2,1,0},{1,0,2}};
+  static const Complex epsilon_sgn[6]= {1,1,1,-1,-1,-1};
```
Comment (aportelli) [inline Grid/qcd/utils/BaryonUtils.h]: Don't duplicate the epsilon values and signs, make it a static member of the `BaryonUtils` class

File: Grid/qcd/utils/BaryonUtils.h [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    std::cout << "GammaB " << (GammaB.g) <<  std::endl;
+ 
+  assert(parity==1 || parity == -1 && "Parity must be +1 or -1");
+
+  GridBase *grid = q1_src.Grid();
+
+  Gamma g4(Gamma::Algebra::GammaT); //needed for parity P_\pm = 0.5*(1 \pm \gamma_4)
+
+  static const int epsilon[6][3] = {{0,1,2},{1,2,0},{2,0,1},{0,2,1},{2,1,0},{1,0,2}};
+  static const int epsilon_sgn[6]= {1,1,1,-1,-1,-1};
```
Comment (aportelli) [inline Grid/qcd/utils/BaryonUtils.h]: Duplicate

File: Grid/qcd/utils/BaryonUtils.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+						 const int parity,
+						 robj &result)
+{
+ 
+  assert(parity==1 || parity == -1 && "Parity must be +1 or -1");
+
+  Gamma g4(Gamma::Algebra::GammaT); //needed for parity P_\pm = 0.5*(1 \pm \gamma_4)
+
+  static const int epsilon[6][3] = {{0,1,2},{1,2,0},{2,0,1},{0,2,1},{2,1,0},{1,0,2}};
+  static const int epsilon_sgn[6]= {1,1,1,-1,-1,-1};
```
Comment (aportelli) [inline Grid/qcd/utils/BaryonUtils.h]: Duplicate

File: Grid/qcd/utils/BaryonUtils.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+ See the full license in the file "LICENSE" in the top level distribution directory
+ *************************************************************************************/
+/*  END LEGAL */
+#pragma once
+//#include <Grid/Hadrons/Global.hpp>
+#include <Grid/Eigen/unsupported/CXX11/Tensor>
+
+NAMESPACE_BEGIN(Grid);
+
+#undef DELTA_F_EQ_2
```
Comment (aportelli) [inline Grid/qcd/utils/BaryonUtils.h]: ?

File: Hadrons/Modules/MContraction/Baryon.hpp [hunk 5]
```diff
@@ -58,11 +65,13 @@ class TBaryon: public Module<BaryonPar>
...
     FERM_TYPE_ALIASES(FImpl2, 2);
     FERM_TYPE_ALIASES(FImpl3, 3);
+    BASIC_TYPE_ALIASES(ScalarImplCR, Scalar);
+    SINK_TYPE_ALIASES(Scalar);
     class Result: Serializable
     {
     public:
         GRID_SERIALIZABLE_CLASS_MEMBERS(Result,
-                                        std::vector<std::vector<std::vector<Complex>>>, corr);
+                                        std::vector<Complex>, corr);
```
Comment (aportelli) [inline Hadrons/Modules/MContraction/Baryon.hpp]: Please add metadata in the result allowing to know the operators offline

File: Grid/qcd/utils/BaryonUtils.h (hunk 4 above)
Comment (felixerben) [reply Grid/qcd/utils/BaryonUtils.h]: copy from A2Autils.h - I did not question why it is there. Should I remove it?

File: Hadrons/Modules/MContraction/Baryon.hpp [hunk 6]
```diff
@@ -135,23 +135,49 @@ void TBaryon<FImpl1, FImpl2,FImpl3>::parseGammaString(std::vector<GammaABPair> &
...
+    gammaString = regex_replace(gammaString, std::regex("j12_alt2"),"(Identity GammaYGamma5)");
+    
+    //A single gamma matrix 
+    std::regex rex_g("([0-9a-zA-Z]+)");
+    //The full string we expect
+    std::regex rex("( *\\(( *\\(([0-9a-zA-Z]+) +([0-9a-zA-Z]+) *\\)){2} *\\) *)+");
+    std::smatch sm;
+    std::regex_match(gammaString, sm, rex);
+    assert(sm[0].matched && "invalid gamma structure.");
+
```
Comment (felixerben) [inline Hadrons/Modules/MContraction/Baryon.hpp]: regex_match does not allow to iterate through tokens - it only finds the first matches of each regular expression! this is why I used sregex_iterator afterwards

File: Hadrons/Modules/MContraction/Baryon.hpp [hunk 7]
```diff
@@ -135,23 +135,49 @@ void TBaryon<FImpl1, FImpl2,FImpl3>::parseGammaString(std::vector<GammaABPair> &
...
+    auto gamma_begin = std::sregex_iterator(gammaString.begin(), gammaString.end(), rex_g);
+    auto gamma_end = std::sregex_iterator();
+
+    //couldn't find out how to count the size in the iterator, other than looping through it...
+    int nGamma=0;
+    for (std::sregex_iterator i = gamma_begin; i != gamma_end; ++i) {
+	nGamma++;
+    }   
+    gammaList.resize(nGamma/4);
```
Comment (felixerben) [inline Hadrons/Modules/MContraction/Baryon.hpp]: there must be some kind of .size() object, but I could not find it

File: Hadrons/Modules/MContraction/Baryon.hpp [hunk 8]
```diff
@@ -135,23 +135,49 @@ void TBaryon<FImpl1, FImpl2,FImpl3>::parseGammaString(std::vector<GammaABPair> &
...
+    }
+    for (int i = 0; i < gammaList.size(); i++){
+	std::vector<Gamma::Algebra> gS1 = strToVec<Gamma::Algebra>(gS[4*i]);
+	std::vector<Gamma::Algebra> gS2 = strToVec<Gamma::Algebra>(gS[4*i+1]);
+	std::vector<Gamma::Algebra> gS3 = strToVec<Gamma::Algebra>(gS[4*i+2]);
+	std::vector<Gamma::Algebra> gS4 = strToVec<Gamma::Algebra>(gS[4*i+3]);
+        gammaList[i].first.first=gS1[0];
+        gammaList[i].first.second=gS2[0];
+        gammaList[i].second.first=gS3[0];
+        gammaList[i].second.second=gS4[0];
```
Comment (felixerben) [inline Hadrons/Modules/MContraction/Baryon.hpp]: extremely ugly, there must be a better way

File: Hadrons/Modules/MContraction/Baryon.hpp [hunk 9]
```diff
@@ -149,29 +184,31 @@ void TBaryon<FImpl1, FImpl2, FImpl3>::execute(void)
...
+    {
+        result[i].gammaA_left = gammaList[i].first.first;
+        result[i].gammaA_left = gammaList[i].first.first;
+        result[i].gammaB_right = gammaList[i].second.second;
+        result[i].gammaB_right = gammaList[i].second.second;
+        result[i].corr.resize(nt);
+        result[i].parity = parity;
+        result[i].quarks = par().quarks;
+        result[i].prefactors = par().prefactors;
+    }
```
Comment (felixerben) [inline Hadrons/Modules/MContraction/Baryon.hpp]: do we want ore metdata here?

**Comment by felixerben** [comment]
The Result object now has the standardised structure, a correlator with Metadata. I also used std::distance to get the size of the regex_iterator.

One last thing we can think of adding: Test_hadrons_spectrum.cc only tests one of the three cases (where the Propagators are not SlicedPropagators and with and MSink module). Do we want to add the other two cases to the test?

Output checked and it looks fine.

**Comment by felixerben** [comment]
Documentation for the latest version is here: https://aportelli.github.io/Hadrons-doc/#/mcontraction

**Pull request merged**